packages/eslint-plugin/src/rules/require-types-exports.ts

```diff
@@ -0,0 +1,386 @@
+import {
+  DefinitionType,
+  ImplicitLibVariable,
+  ScopeType,
+} from '@typescript-eslint/scope-manager';
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES, simpleTraverse } from '@typescript-eslint/utils';
+
+import { createRule, findVariable } from '../util';
+
+type MessageIds = 'requireTypeExport';
+
+export default createRule<[], MessageIds>({
+  name: 'require-types-exports',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Require exporting types that are used in exported entities',
+      recommended: 'strict',
+    },
+    messages: {
+      requireTypeExport: 'Expected type "{{ name }}" to be exported',
```

Comment:
[Docs] What do you think of this, for a more informative error?

```suggestion
      requireTypeExport: '"{{ name }}" is used in other exports from this file, so it should also be exported.',
```

Reply:
I like it
Done